src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+  Atom atom -> summary atom
+  expr -> unknown expr
+
+summary :: Atom n -> OCCM n (IxExpr n)
+summary atom = case atom of
+  Var name -> ixExpr name
+  Lam _ -> invalid "Lam"
+  Con c -> constructor c
+  ACase _ _ _ -> error "Unexpected ACase outside of Simplify"
+  _ -> unknown atom
+  where
+    invalid tag = error $ "Unexpected indexing by " ++ tag
+    constructor = \case
+      -- TODO Represent the actual literal value?
+      Lit _ -> return $ Deterministic []
+      ProdCon elts -> Product <$> mapM summary elts
+      SumCon _ tag payload -> Inject tag <$> summary payload
+      SumAsProd _ (TagRepVal tag) payloads ->
+        Inject (fromIntegral tag) <$> (summary $ payloads !! (fromIntegral tag))
+      SumAsProd _ _ _ -> unknown atom
+      LabelCon _ -> invalid "LabelCon"
+      Newtype _ e -> summary e
+      BaseTypeRef _ -> invalid "BaseTypeRef"
+      TabRef _ -> invalid "TabRef"
+      ConRef _ -> invalid "ConRef"
+      ExplicitDict _ _ -> invalid "ExplicitDict"
+      DictHole _ _ -> invalid "DictHole"
+
+unknown :: HoistableE e => e n -> OCCM n (IxExpr n)
+unknown obj = do
+  freeIxs <- mapM ixExpr $ freeAtomVarsList obj
+  let freeIterationNames = nub $ concatMap freeAtomVarsList freeIxs
+  return $ Deterministic $ freeIterationNames
+
+-- === The actual occurrence analysis ===
+
+class HasOCC (e::E) where
+  occ :: Access n -> e n -> OCCM n (e n)
+  default occ :: (GenericE e, HasOCC (RepE e)) => Access n -> e n -> OCCM n (e n)
+  occ a e = confuseGHC >>= \_ -> toE <$> occ a (fromE e)
+  {-# INLINE occ #-}
+
+instance HasOCC (Name AtomNameC) where
+  occ a n = modify (<> FV (singletonNameMapE n $ AccessInfo 1 a)) $> n
+  {-# INLINE occ #-}
+instance HasOCC (Name HandlerNameC ) where occ _ n = return n
+instance HasOCC (Name DataDefNameC ) where occ _ n = return n
+instance HasOCC (Name EffectNameC  ) where occ _ n = return n
+instance HasOCC (Name InstanceNameC) where occ _ n = return n
+instance HasOCC (Name ClassNameC   ) where occ _ n = return n
+instance HasOCC (Name SpecializedDictNameC) where occ _ n = return n
+
+instance HasOCC Block where
+  occ a (Block ann decls ans) = case (ann, decls) of
+    (NoBlockAnn      , Empty) -> Block NoBlockAnn Empty <$> occ a ans
+    (NoBlockAnn      , _    ) -> error "should be unreachable"
+    (BlockAnn ty effs, _    ) -> do
+      -- The free vars accumulated in the state of OCCM should
+      -- correspond to the free vars of the Abs of the block answer,
+      -- by the decls traversed so far. occNest takes care to uphold
+      -- this invariant, but we temporarily reset the state to an
+      -- empty map, just so that names from the surrounding block
+      -- don't end up influencing elimination decisions here. Note
+      -- that we restore the state (and accumulate free vars of the
+      -- analyzed block into it) right after occNest.
+      Abs decls' ans' <- locally $ occNest a decls ans
+      ty' <- occ accessOnce ty  -- TODO Is free variable counting enough here?
```

Comment:
Say more?  I wasn't sure the right treatment of occurrences in type annotations.  Can a computed array appear in a type?  Do we need to take care to not inline it if it does?  Or is that a different kind of "think twice about inlining this" than occurrences outside of types?

Reply:
> Can a computed array appear in a type?

Yes. It happens in the sets-as-indices example to name one.

> Do we need to take care to not inline it if it does?

You should preserve the binding I think, but it should not affect the other inlining decisions. We need the binding to be able to decide type equality, but if it's only used in types, then ideally we shouldn't even generate code for it... I guess it might require some more thought even outside of this pass.

Reply:
In the meantime, how about creating a helper function that deals with traversing types? I see this pattern appearing multiple times, and in the future it is likely that we will want to do something else.

Reply:
Done.